rosdiagnostic/src/rosdiagnostic/rosdiagnostic.py

```diff
@@ -0,0 +1,234 @@
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+import re
+import os
+import sys
+import rospy
+import rospkg
+import rosbag
+import rosgraph
+import argparse
+import socket
+from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus
+from datetime import datetime
+from rospkg.common import PACKAGE_FILE
+from rospkg.rospack import ManifestManager
+from dateutil.tz import tzlocal
+
+
+class ROSTopicException(Exception):
+    """
+    Base exception class of rostopic-related errors
+    """
+    pass
+
+
+class ROSTopicIOException(ROSTopicException):
+    """
+    rostopic errors related to network I/O failures
+    """
+    pass
+
+
+def _check_master():
+    """
+    Make sure that master is available
+    :raises: :exc:`ROSTopicException` If unable to successfully communicate with master
+    """
+    try:
+        rosgraph.Master('/rostopic').getPid()
+    except socket.error:
+        raise ROSTopicIOException("Unable to communicate with master!")
+
+
+class CprRosDiagnostics(object):
+
+    DIAG_MSG_LVL_STR_NOCOLOR = {
+        DiagnosticStatus.OK:    ' OK  ',
+        DiagnosticStatus.WARN:  'WARN ',
+        DiagnosticStatus.ERROR: 'ERROR',
+        DiagnosticStatus.STALE: 'STALE',
+    }
+
+    DIAG_MSG_LVL_STR_COLOR = {
+        DiagnosticStatus.OK:    '\033[92m OK  \033[0m',
+        DiagnosticStatus.WARN:  '\033[93mWARN \033[0m',
+        DiagnosticStatus.ERROR: '\033[91mERROR\033[0m',
+        DiagnosticStatus.STALE: '\033[94mSTALE\033[0m',
+    }
+
+    def __init__(self, options):
+        self._lvl_strings_map = self.DIAG_MSG_LVL_STR_NOCOLOR if options.nocolor else self.DIAG_MSG_LVL_STR_COLOR
+        self._options = options
+        rospy.Subscriber('{}/diagnostics_agg'.format(options.ns), DiagnosticArray, self._diag_callback, queue_size=1)
+
+    def _get_ns(self, name):
+        # Split namespace by slashes, return the base namespace.
+        return '/'.join(name.split('/')[:-1])
+
+    def _get_non_leaf_statuses(self, statuses):
+        return [self._get_ns(s.name) for s in statuses]
+
+    def _get_leaf_statuses(self, statuses):
+        parent_namespaces = self._get_non_leaf_statuses(statuses)
+        return [s for s in statuses if s.name not in parent_namespaces]
+
+    def _print_begin_banner(self, msg):
+        dt = datetime.fromtimestamp(msg.header.stamp.to_sec(), tzlocal())
+        print('=====================================================================')
+        print('Diagnostic generated on: {}'.format(dt))
+        print('---------------------------------------------------------------------')
```

Comment:
Empy is a templating scheme used in a bunch of core ROS stuff. Simple example:

https://github.com/mikepurvis/catkin_tools_python/blob/aeceac7c4bdeb5577e02c749b81a4c1a97e59bfc/catkin_tools_python/create_python_package_xmls.py#L21-L37